src/nvim/mouse.c

```diff
@@ -412,6 +412,18 @@ bool do_mouse(oparg_T *oap, int c, int dir, int count, bool fixindent)
 
   // Middle mouse button does a 'put' of the selected text
   if (which_button == MOUSE_MIDDLE) {
+    // handled outside any specific mode
+    if (tab_page_click_defs != NULL) {  // only when initialized
+      // Check for clicking in the tab page line.
+      if (mouse_grid <= 1 && mouse_row == 0 && firstwin->w_winrow > 0) {
```

Comment:
can/should this live in the other block that checks this same condition(s), on line 476 (new: 488)

Reply:
why not move this code: https://github.com/neovim/neovim/blob/ce5a9bfe7e537c81d34bd4a27fc6638f20114e67/src/nvim/mouse.c#L413-L414

_after_ this code:
https://github.com/neovim/neovim/blob/ce5a9bfe7e537c81d34bd4a27fc6638f20114e67/src/nvim/mouse.c#L476

and put the middle-click handling in there? special-casing stuff in random places makes the logic very hard to follow.